### ISSUE TITLE
[tflite conversion] Output order changes with multiple outputs

### ISSUE THREAD

**Issue by seanshpark** [tensorflow/tensorflow#51858]
I maybe wrong... is so, could anyone provide how to fix this?

### 1. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installation (pip package or built from source): pip package version 2.6.0
- TensorFlow library (version, if pip package or github SHA, if built from source): N/A

### 2. Code

```py
import tensorflow as tf

print(tf.version.VERSION)

input = tf.keras.layers.Input(shape=(3,3,32), name="input")

o1 = tf.keras.layers.Conv2D(2, (1,1), activation='relu', input_shape=(1,3,3,32))(input)
o2 = tf.keras.layers.Conv2D(16, (1,1), activation='relu', input_shape=(1,3,3,32))(input)
o3 = tf.keras.layers.Conv2D(32, (1,1), activation='relu', input_shape=(1,3,3,32))(input)

model = tf.keras.Model(inputs=input, outputs=[o1,o2,o3])
model.summary()

tf.keras.models.save_model(model, "test_saved_model")
```

```py
import tensorflow as tf

print(tf.version.VERSION)

converter = tf.lite.TFLiteConverter.from_saved_model("test_saved_model")
converter.allow_custom_ops = True
converter.experimental_new_converter = True
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS]

tflite_model = converter.convert()
open("test_saved_model.tflite", "wb").write(tflite_model)
```

Provide code to help us reproduce your issues using one of the following options:

### 3. Failure after conversion
conversion is successful, but the generated model is wrong;

- Output order has changed (order of last axis: 2, 32, 16)

![image](https://user-images.githubusercontent.com/4616940/132270784-01d0bfcf-c036-4cd9-9cd7-98eea93fd075.png)

- Expected something like this (order of last axis: 2, 16, 32)
![image](https://user-images.githubusercontent.com/4616940/132270636-fdff902f-c468-4006-abd8-58a58f0df919.png)

**Comment by abattery** [comment]
Please consider using the signature instead of relying on the tensor indices.
https://www.tensorflow.org/lite/guide/signatures

**Comment by seanshpark** [comment]
> Please consider using the signature instead of relying on the tensor indices.

Thanks for point this out.
I think need some time to understand how to apply to my situation.

**Comment by seanshpark** [comment]
@abattery , if possible, could you please apply this signature to my above examples?
I use `tflite` file as input to another projects but not the interpreter.

**Comment by abattery** [comment]
In such cases, you can read the signature information, stored in the flatbuffer schema:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/schema/schema.fbs#L1193

**Comment by abattery** [comment]
FYI, the signature information will be automatically carried from the TensorFlow 2.5 version if you are using the `from_saved_model` path.

**Comment by seanshpark** [comment]
> In such cases, you can read the signature information from the TensorFlow Lite model file, stored in the flatbuffer schema:
> FYI, the signature information will be automatically carried into the TensorFlow Lite flatbuffer file since the TensorFlow 2.5 version if you are using the from_saved_model path.

Thanks for the information!
We're currently based on tflite schema from TF 2.3.0, so `SignatureDef` doesn't exist.
But will upgrade to latest version soon... found a good reason for this :)

**Comment by gideonKogan** [comment]
> Please consider using the signature instead of relying on the tensor indices. https://www.tensorflow.org/lite/guide/signatures

I have reviewed the suggested method but it does not seem to work. Seems that the input/output order in the signatures is always ordered, unlike in the model itself. Would you please provide an example code? in my code:

```
import tensorflow as tf

print(tf.version.VERSION)

input = tf.keras.layers.Input(shape=(3,3,32), name="input")

o1 = tf.keras.layers.Conv2D(2, (1,1), activation='relu', input_shape=(1,3,3,32))(input)
o2 = tf.keras.layers.Conv2D(16, (1,1), activation='relu', input_shape=(1,3,3,32))(input)
o3 = tf.keras.layers.Conv2D(32, (1,1), activation='relu', input_shape=(1,3,3,32))(input)

model = tf.keras.Model(inputs=input, outputs=[o1,o2,o3])
print(model.summary())

tf.keras.models.save_model(model, "test_saved_model")
converter = tf.lite.TFLiteConverter.from_saved_model("test_saved_model")
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS]

tflite_model = converter.convert()
interpreter = tf.lite.Interpreter(model_content=tflite_model)
interpreter.allocate_tensors()
print(interpreter.get_signature_list())
output_details = interpreter.get_output_details()
print(output_details)
```
Leads to the printout that shows the disagreement between the names and the signatures and the output order, based on the shapes of the outputs:
```
2.10.0
Model: "model_1"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
==================================================================================================
 input (InputLayer)             [(None, 3, 3, 32)]   0           []                               
                                                                                                  
 conv2d_3 (Conv2D)              (None, 3, 3, 2)      66          ['input[0][0]']                  
                                                                                                  
 conv2d_4 (Conv2D)              (None, 3, 3, 16)     528         ['input[0][0]']                  
                                                                                                  
 conv2d_5 (Conv2D)              (None, 3, 3, 32)     1056        ['input[0][0]']                  
                                                                                                  
==================================================================================================
Total params: 1,650
Trainable params: 1,650
Non-trainable params: 0
__________________________________________________________________________________________________
None
WARNING:tensorflow:Compiled the loaded model, but the compiled metrics have yet to be built. `model.compile_metrics` will be empty until you train or evaluate the model.
WARNING:tensorflow:Compiled the loaded model, but the compiled metrics have yet to be built. `model.compile_metrics` will be empty until you train or evaluate the model.
WARNING:absl:Found untraced functions such as _jit_compiled_convolution_op, _jit_compiled_convolution_op, _jit_compiled_convolution_op while saving (showing 3 of 3). These functions will not be directly callable after loading.
INFO:tensorflow:Assets written to: test_saved_model/assets
INFO:tensorflow:Assets written to: test_saved_model/assets
{'serving_default': {'inputs': ['input'], 'outputs': ['conv2d_3', 'conv2d_4', 'conv2d_5']}}
[{'name': 'StatefulPartitionedCall:1', 'index': 8, 'shape': array([ 1,  3,  3, 16], dtype=int32), 'shape_signature': array([-1,  3,  3, 16], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0), 'quantization_parameters': {'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}, 'sparsity_parameters': {}}, {'name': 'StatefulPartitionedCall:2', 'index': 9, 'shape': array([ 1,  3,  3, 32], dtype=int32), 'shape_signature': array([-1,  3,  3, 32], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0), 'quantization_parameters': {'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}, 'sparsity_parameters': {}}, {'name': 'StatefulPartitionedCall:0', 'index': 7, 'shape': array([1, 3, 3, 2], dtype=int32), 'shape_signature': array([-1,  3,  3,  2], dtype=int32), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0), 'quantization_parameters': {'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}, 'sparsity_parameters': {}}]
2022-09-13 16:31:50.943899: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:362] Ignored output_format.
2022-09-13 16:31:50.943920: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:365] Ignored drop_control_dependency.
2022-09-13 16:31:50.944159: I tensorflow/cc/saved_model/reader.cc:45] Reading SavedModel from: test_saved_model
2022-09-13 16:31:50.946248: I tensorflow/cc/saved_model/reader.cc:89] Reading meta graph with tags { serve }
2022-09-13 16:31:50.946270: I tensorflow/cc/saved_model/reader.cc:130] Reading SavedModel debug info (if present) from: test_saved_model
2022-09-13 16:31:50.949384: I tensorflow/cc/saved_model/loader.cc:229] Restoring SavedModel bundle.
2022-09-13 16:31:50.965164: I tensorflow/cc/saved_model/loader.cc:213] Running initialization op on SavedModel bundle at path: test_saved_model
2022-09-13 16:31:50.972867: I tensorflow/cc/saved_model/loader.cc:305] SavedModel load for tags { serve }; Status: success: OK. Took 28714 microseconds.
```

**Comment by seanshpark** [comment]
> This leads to the printout that shows the disagreement between the signatures names and the output order (based on the shapes)

Can you please point out the log where the order is incorrect?

**Comment by seanshpark** [comment]
As this issue is close, how about adding a new issue for your problem? (optionally, with a link to this may help)

**Comment by krunalvaghani** [comment]
Any update, how we can solve the issue? with tensorflow==2.8.0
[Example ](https://www.tensorflow.org/lite/models/modify/model_maker/object_detection)from TensorFlow, with custom data, changes order of the output.

**Comment by AlexanderLutsenko** [comment]
@seanshpark There's a workaround, check it out:
https://github.com/AlexanderLutsenko/tflite_converter_fixer
